src/plugins/rgbcolor/rgbcolor.c

```diff
@@ -0,0 +1,379 @@
+					  { "tomato", 0xff6347ff },
+					  { "turquoise", 0x40e0d0ff },
+					  { "violet", 0xee82eeff },
+					  { "wheat", 0xf5deb3ff },
+					  { "white", 0xffffffff },
+					  { "whitesmoke", 0xf5f5f5ff },
+					  { "yellow", 0xffff00ff },
+					  { "yellowgreen", 0x9acd32ff },
+					  { NULL, 0 } };
+
+typedef enum
+{
+	COLOR_INVALID,
+	HEX_THREE,
+	HEX_FOUR,
+	HEX_SIX,
+	HEX_EIGHT,
+	NAMED_COLOR
+} ColorVariant;
+
+static const NamedColor * findNamedColor (const char * colorName)
+{
+	const NamedColor * cur = &NamedColors[0];
+	while (cur->name != NULL)
+	{
+		if (strcmp (cur->name, colorName) == 0)
+		{
+			return cur;
+		}
+		++cur;
+	}
+	return NULL;
+}
+
+static ColorVariant is_valid_key (Key * key, Key * parentKey)
+{
+	const Key * meta = keyGetMeta (key, "check/rgbcolor");
+	if (!meta) return 1;
+	const char * value = keyString (key);
+	const char * regexString = "^#([0-9a-fA-F]{3,4}|[0-9a-fA-F]{6}|[0-9a-fA-F]{8})$";
+
+	regex_t regex;
+	regmatch_t offsets;
+	int compile_failure = regcomp (&regex, regexString, REG_NOSUB | REG_EXTENDED | REG_NEWLINE);
+	if (compile_failure) return -1;
+	// regexec returns 0 on a successful match
+	int match = !(regexec (&regex, value, 0, &offsets, 0));
+	regfree (&regex);
+
+	const NamedColor * namedColor = findNamedColor (value);
+
+	if (namedColor != NULL)
+	{
+		return NAMED_COLOR;
+	}
+
+	if (!match && namedColor == NULL)
+	{
+		ELEKTRA_SET_ERRORF (214, parentKey, "Key %s with value %s is neither a valid hex formatted color nor a named color.",
+				    keyName (key), keyString (key));
+		return COLOR_INVALID;
+	}
+
+	int len = strlen (value);
+	switch (len)
+	{
+	case 4:
+		return HEX_THREE;
+	case 5:
+		return HEX_FOUR;
+	case 7:
+		return HEX_SIX;
+	case 9:
+		return HEX_EIGHT;
+	default:
+		return COLOR_INVALID;
+	}
+}
+
+static void elektraColorSetInteger (Key * key, kdb_unsigned_long_t c)
+{
+	char colorStr[11];
+	snprintf (colorStr, 11, "%u", c);
+
+	ELEKTRA_LOG_DEBUG ("Set %s to integer %s", keyName (key), colorStr);
+	keySetString (key, colorStr);
+	keySetMeta (key, "type", "unsigned_long");
+}
+
+static char * elektraColorExpand (const char * str, ColorVariant colVar)
+{
+	// Expand #abcd to #aabbccdd or #abc to #aabbccff
+	char * expandedStr = (char *) elektraMalloc (10);
+
+	if (colVar == HEX_THREE || colVar == HEX_FOUR)
+	{
+		expandedStr[0] = '#';
+
+		for (size_t i = 1; i < strlen (str); i++)
+		{
+			expandedStr[2 * i - 1] = str[i];
+			expandedStr[2 * i] = str[i];
+		}
+	}
+	else
+	{
+		strcpy (expandedStr, str);
+	}
+
+	if (colVar == HEX_THREE || colVar == HEX_SIX)
+	{
+		expandedStr[7] = 'f';
+		expandedStr[8] = 'f';
+	}
+
+	expandedStr[9] = '\0';
+
+	return expandedStr;
+}
+
+static void elektraColorNormalizeHexString (Key * key, ColorVariant colVar)
+{
+	const char * str = keyString (key);
+	char * origvalue = elektraStrDup (str);
+
+	kdb_unsigned_long_t color;
+	if (colVar == NAMED_COLOR)
+	{
+		const NamedColor * namedColor = findNamedColor (str);
```

Comment:
If you can somehow find a way to avoid calling `findNamedColor` a second time that would be nice.

Reply:
Is this done now?

Reply:
I haven't found a reasonably simple way to avoid this. I think this would require a rewrite of how the validity of a key is checked and wherever it is called, which is a fairly big part of the plugin. I would leave it as is, if that's fine.

Reply:
Yes, of course it is fine.